[mlir] Improve mlir-query tool by implementing `getBackwardSlice` matcher

chios202 opened the pull request:
Improve mlir-query tool by implementing `getBackwardSlice` matcher

Note: backwardSlice and forwardSlice algoritms are the same as the ones in `mlir/lib/Analysis/SliceAnalysis.cpp`
Example of current matcher. The query was made to the file: `mlir/test/mlir-query/complex-test.mlir`

```mlir
./mlir-query /home/dbudii/personal/llvm-project/mlir/test/mlir-query/complex-test.mlir -c "match getDefinitions(hasOpName(\"arith.add
f\"),2)"

Match #1:

/home/dbudii/personal/llvm-project/mlir/test/mlir-query/complex-test.mlir:5:8:
  %0 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel"]} ins(%arg0 : tensor<5x5xf32>) outs(%arg1 : tensor<5x5xf32>) {
       ^
/home/dbudii/personal/llvm-project/mlir/test/mlir-query/complex-test.mlir:7:10: note: "root" binds here
    %2 = arith.addf %in, %in : f32
         ^
Match #2:

/home/dbudii/personal/llvm-project/mlir/test/mlir-query/complex-test.mlir:10:16:
  %collapsed = tensor.collapse_shape %0 [[0, 1]] : tensor<5x5xf32> into tensor<25xf32>
               ^
/home/dbudii/personal/llvm-project/mlir/test/mlir-query/complex-test.mlir:13:11:
    %c2 = arith.constant 2 : index
          ^
/home/dbudii/personal/llvm-project/mlir/test/mlir-query/complex-test.mlir:14:18:
    %extracted = tensor.extract %collapsed[%c2] : tensor<25xf32>
                 ^
/home/dbudii/personal/llvm-project/mlir/test/mlir-query/complex-test.mlir:15:10: note: "root" binds here
    %2 = arith.addf %extracted, %extracted : f32
         ^
2 matches.
```

jpienaar commented on mlir/include/mlir/IR/Matchers.h:
```diff
@@ -59,15 +59,15 @@ struct NameOpMatcher {
   bool match(Operation *op) { return op->getName().getStringRef() == name; }
 
-  StringRef name;
+  std::string name;
```
Why is this change needed?

chios202 replied:
This change was necessary due to a bug I found while implementing the LetQuery. If one used hasOpName("arith.mulf") it would work just fine, but when used `let myMatch hasOpName("arith.mulf")` there was a dangling reference when the query was insterted into the `namedValues `container

chios202 replied:
I let this for now as it is, even though we are not introducing the LetQuery feature now. I think it's still a good thing to prevent future bugs, but I could revert this back to `StringRef`.

jpienaar commented on mlir/include/mlir/Query/Matcher/ExtraMatchers.h:
```diff
@@ -0,0 +1,193 @@
+//===- Matchers.h - Various common matchers ---------------------*- C++ -*-===//
```
You'd need to update the filename here.

jpienaar commented on mlir/include/mlir/Query/Matcher/ExtraMatchers.h:
```diff
@@ -0,0 +1,193 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides extra matchers that are very useful for mlir-query
```
All the matchers in this dir is for mlir-query.

chios202 replied:
So I think we could remove this description?

jpienaar replied:
Yes, SGTM. Should these even be merged with the others ? (e.g., I'm not sure when a matcher would be considered extra or not)

chios202 replied:
The naming is a bit redundant, I think a more fitting name could be ComplexMatchers or SliceMatchers. Moving them to mlir/include/mlir/IR/Matchers.h could also be a good approach I believe.

jpienaar commented on mlir/include/mlir/Query/Matcher/MatchersInternal.h:
```diff
@@ -1,19 +1,8 @@
 //===- MatchersInternal.h - Structural query framework ----------*- C++ -*-===//
 //
-// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// Part of the LLVM Project, under the Apache License v2.0 wIth LLVM Exceptions.
```
Typo?

jpienaar commented on mlir/lib/Query/Matcher/Parser.cpp:
```diff
@@ -144,6 +156,53 @@ class Parser::CodeTokenizer {
+    while (length < code.size() && isdigit(code[length]))
+      ++length;
+
+    // Try to recognize a floating point literal.
```
Where are floats needed in query?

chios202 replied:
They are not needed. I will remove the float parser part.

jpienaar commented on mlir/lib/Query/Query.cpp:
```diff
@@ -114,6 +123,8 @@ LogicalResult HelpQuery::run(llvm::raw_ostream &os, QuerySession &qs) const {
   os << "Available commands:\n\n"
         "  match MATCHER, m MATCHER      "
         "Match the mlir against the given matcher.\n"
+        "  let NAME MATCHER, l NAME MATCHER  "
```
Could this be split out into it's own change?

chios202 replied:
We could do this

chios202 replied:
I removed this feature

jpienaar reviewed (commented):
Quick skim comments.

dbudii commented on mlir/include/mlir/Query/Matcher/MatchersInternal.h:
```diff
@@ -22,16 +11,45 @@
+#include <memory>
+#include <stack>
+#include <unordered_set>
+#include <vector>
```
Some of these can be removed too

jpienaar commented:
I see the presubmit is failing on the newly added tests (mlir-query/complex-test.mlir failing)

jpienaar commented on mlir/include/mlir/Query/Matcher/ExtraMatchers.h:
```diff
@@ -0,0 +1,188 @@
+//===- ExtraMatchers.h - Various common matchers ---------------------*- C++
```
Nit: make fit on one line

jpienaar commented on mlir/include/mlir/Query/Matcher/ExtraMatchers.h:
```diff
@@ -0,0 +1,188 @@
+  bool matches(Operation *op, SetVector<Operation *> &backwardSlice,
+               QueryOptions &options, unsigned tempHops) {
+
+    if (op->hasTrait<OpTrait::IsIsolatedFromAbove>()) {
```
It would be good here to document behavior of BackwardSliceMatcher. Give a little example with a before and after. Helps with also talking about edge cases (is an op it is own backwardslice?)

chios202 replied:
I added extensive documentation for these matchers and moved them to `Matchers.h` along with the existing ones. As you pointed out, I'm not sure they belonged in the `ExtraMatchers.h` file.

jpienaar commented on mlir/include/mlir/Query/Matcher/ExtraMatchers.h:
```diff
@@ -0,0 +1,188 @@
+      return false;
+    }
+
+    auto processValue = [&](Value value) {
```
Could mlir/include/mlir/Analysis/SliceAnalysis.h's methods be used here?

chios202 replied:
temHops parameter which specifies how far into the graph we are searching (how many levels) limits us to reuse code from mlir/include/mlir/Analysis/SliceAnalysis.h

chios202 replied:
I think `maxDepth` would be more fitting for this parameter.

jpienaar replied:
Looking at it, it may be possible (but less efficient) by keeping track of the depth in a map and then having the filter return whether the minimum depth of any of the operation's argument's depth is greater or equal to maxDepth.

jpienaar commented on mlir/include/mlir/Query/Matcher/ExtraMatchers.h:
```diff
@@ -0,0 +1,188 @@
+
+private:
+  bool matches(Operation *op, SetVector<Operation *> &forwardSlice,
+               QueryOptions &options, unsigned tempHops) {
```
Could you add documentation for the args? tempHops is a bit non-descriptive.

jpienaar commented on mlir/include/mlir/Query/Matcher/ExtraMatchers.h:
```diff
@@ -0,0 +1,188 @@
+  ForwardSliceMatcher(matcher::DynMatcher &&innerMatcher, unsigned hops)
+      : innerMatcher(std::move(innerMatcher)), hops(hops) {}
+
+private:
```
Nit: I prefer keeping public section together followed by private, that way one sees the interface all together.

jpienaar commented on mlir/include/mlir/Query/Matcher/ExtraMatchers.h:
```diff
@@ -0,0 +1,188 @@
+#include "mlir/Query/Query.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace mlir {
```
Use C++17 style nested namespace decl here.

jpienaar commented on mlir/include/mlir/Query/Matcher/MatchFinder.h:
```diff
@@ -7,32 +7,60 @@
-  static std::vector<Operation *> getMatches(Operation *root,
-                                             DynMatcher matcher) {
-    std::vector<Operation *> matches;
+private:
```
Same as above.

jpienaar commented on mlir/include/mlir/Query/Matcher/MatchFinder.h:
```diff
@@ -7,32 +7,60 @@
+  static std::vector<Operation *>
+  getMatches(Operation *root, QueryOptions &options, DynMatcher matcher,
+             llvm::raw_ostream &os, QuerySession &qs) {
+    unsigned matchCount = 0;
```
Lets use int here, prefer to use signed when doing arithmetic (https://groups.google.com/g/llvm-dev/c/q80EgETJGh0/m/hKCh7N4XAgAJ for context why I prefer it). The usage internally isn't consistent unfortunately.

jpienaar commented on mlir/include/mlir/Query/Matcher/MatchFinder.h:
```diff
@@ -7,32 +7,60 @@
+          os << "Match #" << ++matchCount << ":\n\n";
+          SmallVector<Operation *> printingOps(tempStorage.takeVector());
+          for (auto op : printingOps) {
+            printMatch(os, qs, op, "root");
```
Would this result in marking these all as root?

jpienaar commented on mlir/include/mlir/Query/Matcher/MatchersInternal.h:
```diff
@@ -1,4 +1,3 @@
-//===- MatchersInternal.h - Structural query framework ----------*- C++ -*-===//
 //
```
Accidental?

jpienaar commented on mlir/include/mlir/Query/Matcher/MatchersInternal.h:
```diff
@@ -10,28 +9,53 @@
 //
 // Matchers are methods that return a Matcher which provides a method
```
Update comment as there are multiple

jpienaar commented on mlir/include/mlir/Query/Matcher/MatchersInternal.h:
```diff
@@ -10,28 +9,53 @@
+namespace mlir {
+namespace query {
+struct QueryOptions;
+}
```
Missing comment. But also if this is just nested here, use C++17 nested namespace.

chios202 replied:
I omitted this, but I'll include it with any new changes. Otherwise, I'll include it anyway.

jpienaar commented on mlir/include/mlir/Query/Matcher/MatchersInternal.h:
```diff
@@ -10,28 +9,53 @@
+struct has_simple_match : std::false_type {};
+
+template <typename T>
+struct has_simple_match<T, std::void_t<decltype(std::declval<T>().match(
```
Could you document these?

jpienaar commented on mlir/include/mlir/Query/Matcher/VariantValue.h:
```diff
@@ -92,8 +93,15 @@ class VariantValue {
+  // Unsigned value functions.
+  bool isUnsigned() const;
+  unsigned getUnsigned() const;
+  void setUnsigned(unsigned Unsigned);
```
MLIR's naming convention is camelBack (https://mlir.llvm.org/getting_started/DeveloperGuide/#style-guide).

jpienaar commented on mlir/include/mlir/Query/Query.h:
```diff
@@ -17,7 +17,13 @@
+  bool inclusive = true;
+};
+
+enum class QueryKind { Invalid, NoOp, Help, Match, Quit, SetBool };
```
Beyond Invalid, any reason to not have these sorted?

jpienaar commented on mlir/include/mlir/Query/Query.h:
```diff
@@ -17,7 +17,13 @@
 namespace mlir::query {
 
-enum class QueryKind { Invalid, NoOp, Help, Match, Quit };
+struct QueryOptions {
```
This is part of public interface, lets use doxygen comments for.

jpienaar commented on mlir/include/mlir/Query/QuerySession.h:
```diff
@@ -9,14 +9,18 @@
 
+namespace mlir::query::matcher {
+class Registry;
+}
```
Missing comment

jpienaar commented on mlir/include/mlir/Query/QuerySession.h:
```diff
@@ -33,6 +37,11 @@ class QuerySession {
   bool terminate = false;
 
+public:
+  bool omitBlockArguments = false;
```
So these are set per query session?

chios202 replied:
That's how I thought of it. The user sets these per query session. Should we consider a different approach for this?

jpienaar replied:
Yes, coming back with fresh eyes. These affect only one matcher (e.g., inclusive in isolation is difficult to interpret, while specific to backwards analysis its clear). Could these be made args of the matcher?

jpienaar commented on mlir/lib/Query/Matcher/Parser.cpp:
```diff
@@ -144,6 +156,30 @@ class Parser::CodeTokenizer {
     return result;
   }
 
+  void consumeNumberLiteral(TokenInfo *result) {
```
Where is this called?

chios202 replied:
It is called on line 148

jpienaar replied:
I was missing which query kinds were using it, so m_GetDefinitions is one of them.

jpienaar commented on mlir/test/mlir-query/complex-test.mlir:
```diff
@@ -0,0 +1,39 @@
+// CHECK: %[[COLLAPSED:.*]] = tensor.collapse_shape %0 {{\[\[.*\]\]}} : tensor<5x5xf32> into tensor<25xf32>
+// CHECK: %[[EXTRACTED:.*]] = tensor.extract %[[COLLAPSED]][%c2] : tensor<25xf32>
+// CHECK: %[[ADDF2:.*]] = arith.addf %[[EXTRACTED]], %[[EXTRACTED]] : f32
+
```
Remove some of this vertical whitespace.

jpienaar commented on mlir/test/mlir-query/complex-test.mlir:
```diff
@@ -0,0 +1,39 @@
+
+// CHECK: Match #2:
+
+// CHECK: %[[COLLAPSED:.*]] = tensor.collapse_shape %0 {{\[\[.*\]\]}} : tensor<5x5xf32> into tensor<25xf32>
```
You are capturing on the output but not using the captured in operands, avoid hard coding there too.

chios202 commented on mlir/include/mlir/Query/Matcher/MatchFinder.h:
```diff
@@ -7,32 +7,93 @@
+          SmallVector<Operation *> printingOps(tempStorage.takeVector());
+          for (auto op : printingOps) {
+            if (printMatchingOps) {
+              printMatch(os, qs, op, "root");
```
This should use the overloaded print method with no argument for the binding. I accidentally reverted this back to using the old method, but will update it accordingly with the next changes.

jpienaar replied:
SG

jpienaar commented on mlir/include/mlir/Query/Query.h:
```diff
@@ -17,7 +17,31 @@
+  /// When omitBlockArguments is true, the matcher omits traversing
+  /// any block arguments
+  bool omitBlockArguments = false;
+  /// When omitUsesFromAbove is true, the matcher omits
```
Reflow paragraph to avoid extra whitespace

jpienaar commented on mlir/lib/Query/Matcher/Parser.cpp:
```diff
@@ -135,6 +135,18 @@ class Parser::CodeTokenizer {
     case '\'':
       consumeStringLiteral(&result);
       break;
+    case '0':
```
Would this have previously parsed as an identifier?

chios202 replied:
From what I recall, yes. It would have parsed an identifier, resulting in a `TokenKind::InvalidChar`

jpienaar commented on mlir/lib/Query/Matcher/Parser.cpp:
```diff
@@ -144,6 +156,30 @@ class Parser::CodeTokenizer {
   }
 
+  void consumeNumberLiteral(TokenInfo *result) {
+    unsigned length = 1;
```
Wouldn't stringref consumeInteger also work here?

chios202 replied:
It does. Thank you for the suggestion

jpienaar commented on mlir/lib/Query/Matcher/VariantValue.cpp:
```diff
@@ -85,12 +92,27 @@ void VariantValue::reset() {
+bool VariantValue::isSigned() const { return type == ValueType::Signed; }
+
+int64_t VariantValue::getSigned() const {
+  assert(isSigned());
```
I can't see how this could possibly fail.

jpienaar commented on mlir/lib/Query/Matcher/VariantValue.cpp:
```diff
@@ -85,12 +92,27 @@ void VariantValue::reset() {
   type = ValueType::Nothing;
 }
 
+// Unsinged
```
Signed

jpienaar commented on mlir/lib/Query/QueryParser.cpp:
```diff
@@ -147,6 +167,7 @@ QueryRef QueryParser::doParse() {
           .Case("#", ParsedQueryKind::Comment, /*isCompletion=*/false)
           .Case("help", ParsedQueryKind::Help)
           .Case("m", ParsedQueryKind::Match, /*isCompletion=*/false)
+          .Case("set", ParsedQueryKind::Set)
```
Lets keep these in alphabetical order

chios202 commented on mlir/include/mlir/IR/Matchers.h:
```diff
@@ -363,8 +366,267 @@ struct RecursivePatternMatcher {
+///    {1, 2, 5, 3, 4, 6}
+///
+
+class BackwardSliceMatcher {
```
Yes, implementations would be better in a .cpp file

chios202 commented on mlir/include/mlir/IR/Matchers.h:
```diff
@@ -363,8 +366,267 @@ struct RecursivePatternMatcher {
+          if (parentOp->getNumRegions() != 1 &&
+              parentOp->getRegion(0).getBlocks().size() != 1) {
+            llvm::errs()
+                << "Error: Expected parentOp to have exactly one region and "
```
This limitation comes from the original implementation of the algoritm; it's treated with an assert there (the parent op must have exactly one region with exactly one block). E.g: I think it would be tricky to figure out whether the other blocks are part of the slice (the flow into this block). I think it would be more fitting to treat this with an assert here too?

chios202 commented on mlir/include/mlir/Query/Matcher/MatchFinder.h:
```diff
@@ -7,32 +7,93 @@
+  //
+  // getMatches walks the IR and prints operations as soon as it matches them
+  // if a matcher is to be further extracted into the function, then it does not
+  // print operations
```
Would we want to return a vector of vectors (each one containing rootOp + matching environment)  for printing and for reusability we could return a vector containing sequences of rootOp + matching environment?

chios202 replied:
It now follows a similar design to clang-query's MatchFinder class

chios202 commented:
A simple implementation where we would reuse most of the original logic while preserving a DFS approach would be: 
```cpp
llvm::DenseMap<Operation *, int64_t> opDepths;
    opDepths[rootOp] = 0;

    BackwardSliceOptions depthOptions = options;
    depthOptions.filter = [&](Operation *op) {
      for (auto operand : op->getOperands()) {
        if (auto definingOp = operand.getDefiningOp()) {
          if (!opDepths.count(definingOp)) {
            if (opDepths[op] > limitDepth) {
              return false;
            }
            opDepths[definingOp] = opDepths[op] + 1;
          }
        } else if (auto blockArgument = dyn_cast<BlockArgument>(operand)) {
          Operation *parentOp = blockArgument.getOwner()->getParentOp();
          if (!opDepths.count(parentOp)) {
            if (opDepths[op] > limitDepth) {
              return false;
            }
            opDepths[parentOp] = opDepths[op] + 1;
          }
        }
      }
      return true;
    };
    getBackwardSlice(rootOp, &backwardSlice, depthOptions);

chios202 commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -74,7 +74,7 @@ static OpFoldResult getDimValue(OpBuilder &builder, Location loc, Value v,
 /// `source`.
-static Operation *getSlice(OpBuilder &b, Location loc, Value source,
+static Operation *getSubviewOrSlice(OpBuilder &b, Location loc, Value source,
                            ArrayRef<OpFoldResult> offsets,
```
I renamed this to avoid naming collision with SliceAnalysis::getSlice

jpienaar replied:
Why is a naming conflict here a problem?

chios202 commented on mlir/lib/Query/Query.cpp:
```diff
@@ -26,15 +27,6 @@ complete(llvm::StringRef line, size_t pos, const QuerySession &qs) {
-
 // TODO: Extract into a helper function that can be reused outside query
 // context.
 static Operation *extractFunction(std::vector<Operation *> &ops,
```
Could this method be a candidate for the MatchFinder class?

jpienaar replied:
Yes, or even higher, such as a utility function in Analysis or Transform or so, as this is rather common.

chios202 replied:
I could move it to MatchFinder. I could also create an issue for this and assign it to myself. I’d need to investigate where in Analysis or Transform would be the most suitable place.

chios202 commented on mlir/lib/IR/Matchers.cpp:
```diff
@@ -0,0 +1,57 @@
+//===- Matchers.h - Various common matchers ---------------------*- C++ -*-===//
```
This should be Matchers.cpp

chios202 commented on mlir/include/mlir/Query/Matcher/MatchFinder.h:
```diff
@@ -7,33 +7,48 @@
-
-    return matches;
-  }
+  /// A subclass which preserves the matching information
```
I do think this class promotes modularity and aligns well with clang-query. I think more features could be incorporated that may be used in or outside the Query context. E.g : we could implement a method, addMatchers(Matcher, Callback), then once a matcher has returned, it's callback for can be invoked. We could also add a subclass, MatcherOptions (this could possibly include the QueryOptions)

jpienaar commented on mlir/include/mlir/IR/Matchers.h:
```diff
@@ -363,8 +364,72 @@ struct RecursivePatternMatcher {
+  bool match(Operation *op, SetVector<Operation *> &backwardSlice,
+             query::QueryOptions &options) {
+
+    if (innerMatcher.match(op) &&
```
Just do

return innerMatcher.match(op) && matches(op, backwardSlice, options, maxDepth);

jpienaar commented on mlir/include/mlir/Query/Matcher/MatchersInternal.h:
```diff
@@ -20,18 +21,50 @@
 #define MLIR_TOOLS_MLIRQUERY_MATCHER_MATCHERSINTERNAL_H
 
-#include "mlir/IR/Matchers.h"
+#include "mlir/Analysis/SliceAnalysis.h"
```
Could this be in a different file? (have internal only be matcher internal details independent of any matchers)

jpienaar replied:
(I don't see any use here TBH)

jpienaar commented on mlir/include/mlir/Query/Matcher/VariantValue.h:
```diff
@@ -21,7 +21,7 @@
 
 // All types that VariantValue can contain.
-enum class ArgKind { Matcher, String };
+enum class ArgKind { Matcher, String, Signed };
```
Lets keep these sorted

jpienaar commented on mlir/include/mlir/Query/QuerySession.h:
```diff
@@ -9,14 +9,14 @@
 #ifndef MLIR_TOOLS_MLIRQUERY_QUERYSESSION_H
 #define MLIR_TOOLS_MLIRQUERY_QUERYSESSION_H
 
+#include "Matcher/VariantValue.h"
```
Lets add prefix (follow line 14)

jpienaar commented on mlir/lib/IR/Matchers.cpp:
```diff
@@ -0,0 +1,59 @@
+//===----------------------------------------------------------------------===//
+
+#include "mlir/IR/Matchers.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
```
We can't have IR depend on Tensor either.

jpienaar replied:
(but don't see need for it here)

jpienaar commented on mlir/lib/IR/Matchers.cpp:
```diff
@@ -0,0 +1,59 @@
+      if (auto definingOp = operand.getDefiningOp()) {
+        // If the defining operation is already in the map, its depth has been
+        // computed; recomputation can be skipped
+        if (!opDepths.contains(definingOp)) {
```
Wouldn't it be the min of all the defining op's depth?

chios202 replied:
Good point

jpienaar commented on mlir/lib/Query/Matcher/MatchFinder.cpp:
```diff
@@ -0,0 +1,69 @@
+                             Operation *op) const {
+  auto fileLoc = dyn_cast<FileLineColLoc>(op->getLoc());
+  SMLoc smloc = qs.getSourceManager().FindLocForLineAndColumn(
+      qs.getBufferId(), fileLoc.getLine(), fileLoc.getColumn());
```
You dyn\_cast but then query without checking if null. Either cast (so that at least one has an assert in non-opt builds) or check.

jpienaar commented on mlir/lib/Query/Matcher/MatchFinder.cpp:
```diff
@@ -0,0 +1,69 @@
+                             Operation *op) const {
+  auto fileLoc = dyn_cast<FileLineColLoc>(op->getLoc());
+  SMLoc smloc = qs.getSourceManager().FindLocForLineAndColumn(
+      qs.getBufferId(), fileLoc.getLine(), fileLoc.getColumn());
```
You dyn\_cast but then query without checking if null. Either cast (so that at least one has an assert in non-opt builds) or check.

chios202 replied:
I did cast. The reasoning behind it being that, ideally this the file location should always be available?

jpienaar commented on mlir/lib/Query/Matcher/VariantValue.cpp:
```diff
@@ -85,12 +92,24 @@ void VariantValue::reset() {
+int64_t VariantValue::getSigned() const { return value.Signed; }
+
+void VariantValue::setSigned(int64_t newValue) {
+  reset();
```
Is this needed?

jpienaar reviewed (commented):
Is the example output in the PR description still the current output?

chios202 committed:
Compute shortest depth in backwardSlice method
Relocate backwardSlice matcher to Query specific headers
Remove unncecessary code

chios202 commented:
I'm not sure why complex-test.mlir is failing in CI for Linux. Locally I use wsl and it seems to pass. Also I did not modify the printing logic.

chios202 commented on mlir/include/mlir/IR/Matchers.h:
```diff
@@ -19,7 +19,8 @@
 #include "mlir/IR/OpDefinition.h"
 #include "mlir/Interfaces/InferIntRangeInterface.h"
-
+#include "mlir/Query/Matcher/MatchersInternal.h"
```
Looking at it again, I think it may be possible to make `BackwardSlice` matcher not to depend on `MatcherInternals` , but I'm not so sure about the dependency on Analysis, I think it may be possible with the use of a function pointer.

chios202 committed:
Update grammar
Make BackwardSlice matcher more generic
Capture values in tests

chios202 committed:
Improve depth limiting approach

chios202 commented:
@jpienaar could you provide feedback on the latest changes?

chios202 commented on mlir/test/mlir-query/complex-test.mlir:
```diff
@@ -0,0 +1,32 @@
+// RUN: mlir-query %s -c "m getAllDefinitions(hasOpName(\"arith.addf\"),2)" | FileCheck %s
+
+#map = affine_map<(d0, d1) -> (d0, d1)>
+func.func @slice_use_from_above(%arg0: tensor<5x5xf32>, %arg1: tensor<5x5xf32>) {
```
I've copied this test from `mlir/test/IR/slice.mlir`  I coud at least rename it to "slicing-query" and/or change the contents to avoid duplication?

jpienaar reviewed (commented):
I think this look overall good. Thanks. Lets do the one rename you were also thinking about, and then can approve. Do you need help landing it?

chios202 committed:
Rename slicing matchers file to SliceMatchers.h
